diff/parse.go

```diff
@@ -426,7 +518,39 @@ func handleEmpty(fd *FileDiff) (wasEmpty bool) {
 		fd.NewName = "/dev/null"
 	}
 
-	return true
+	// For ambiguous 'diff --git' lines, try to reconstruct filenames using extended headers.
+	if success && (isCopy || isRename) && fd.OrigName == "" && fd.NewName == "" {
+		diffArgs := fd.Extended[0][len("diff --git "):]
+
+		reconstruct := func(header string, prefix string, whichFile int, result *string) bool {
+			if !strings.HasPrefix(header, prefix) {
+				return false
+			}
+			rawFilename := header[len(prefix):]
+
+			// extract the filename prefix (e.g. "a/") from the 'diff --git' line.
+			var prefixLetterIndex int
+			if whichFile == 1 {
+				prefixLetterIndex = 0
+			} else if whichFile == 2 {
+				prefixLetterIndex = len(diffArgs) - len(rawFilename) - 2
+			}
+			if prefixLetterIndex < 0 || diffArgs[prefixLetterIndex+1] != '/' {
+				return false
+			}
+
+			*result = diffArgs[prefixLetterIndex:prefixLetterIndex+2] + rawFilename
+			return true
+		}
+
+		for _, header := range fd.Extended {
+			_ = reconstruct(header, "copy from ", 1, &fd.OrigName) ||
+				reconstruct(header, "copy to ", 2, &fd.NewName) ||
+				reconstruct(header, "rename from ", 1, &fd.OrigName) ||
+				reconstruct(header, "rename to ", 2, &fd.NewName)
+		}
```

Comment:
Just an idea, but maybe something like this might be more straightforward?

```suggestion
		for _, header := range fd.Extended {
			switch {
			case strings.HasPrefix(header, "copy from "):
				reconstruct(header[len("copy from "):], 1, &fd.OrigName)
			case strings.HasPrefix(header, "copy to "):
				reconstruct(header[len("copy to "):], 2, &fd.NewName)
			case strings.HasPrefix(header, "rename from "):
				reconstruct(header[len("rename from "):], 1, &fd.OrigName)
			case strings.HasPrefix(header, "rename to "):
				reconstruct(header[len("rename to "):], 2, &fd.NewName)
			}
		}

```